app/controllers/following/posts_controller.rb

```diff
@@ -0,0 +1,5 @@
+class Following::PostsController < ApplicationController
+  def index
+    @posts = Post.where(user: current_user.following)
```

Comment:
id の配列だけでよければ、こっちが効率的です！
```suggestion
    @posts = Post.where(user: current_user.following_ids)
```

Reply:
そうなんですね、has_manyで関連づけたものに使用できる動的なメソッドみたいですね！
具体的な違いはわからないのですが、処理がシンプルなように見えますね。
次のrails勉強会の時にお聞きしたいです！

```
>> Post.where(user: User.first.following)
  User Load (1.8ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
  Post Load (2.9ms)  SELECT "posts".* FROM "posts" WHERE "posts"."user_id" IN (SELECT "users"."id" FROM "users" INNER JOIN "relationships" ON "users"."id" = "relationships"."followed_id" WHERE "relationships"."follower_id" = $1) /* loading for pp */ LIMIT $2
```

```
  >> Post.where(user: User.first.following_ids)
  User Load (0.9ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
  User Pluck (0.4ms)  SELECT "users"."id" FROM "users" INNER JOIN "relationships" ON "users"."id" = "relationships"."followed_id" WHERE "relationships"."follower_id" = $1
```

Reply:
rails console で２つの違いをみてみるとよいかも。
- `user.following`
- `user.following_ids`

Reply:
sqlを実行する時の出力される内容に違いがあることを確認できました。
following_idsの方はUser Pluck。

rails ガイド↓
selectと異なり、pluckはデータベースから受け取った結果を直接Rubyの配列に変換してくれます。そのためのActiveRecordオブジェクトを事前に構成しておく必要はありません。従って、このメソッドは大規模なクエリや利用頻度の高いクエリで使うとパフォーマンスが向上します。ただし、オーバーライドを行なうモデルメソッドは使えません。以下に例を示します。

```
>> User.first.following
  User Load (1.2ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
  User Load (0.8ms)  SELECT "users".* FROM "users" INNER JOIN "relationships" ON "users"."id" = "relationships"."followed_id" WHERE "relationships"."follower_id" = $1 /* loading for pp */ LIMIT $2  [["follower_id", 9], ["LIMIT", 11]]
=> []
>> User.first.following_ids
  User Load (0.9ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
  User Pluck (0.5ms)  SELECT "users"."id" FROM "users" INNER JOIN "relationships" ON "users"."id" = "relationships"."followed_id" WHERE "relationships"."follower_id" = $1  [["follower_id", 9]]
=> []
```

Reply:
イマイチわかりません。。

Reply:
SELECT 文のところが、`SELECT "users".*` から `SELECT "users"."id"` に変化してるので、userのカラムをすべて取得しない方法になってるので効率がいい。

Reply:
あ、そこか。見落としてました。
参照するカラムを減らすことでも効率化できるんですね。
「sql文の効率化 = レコードを参照する数を減らす」だと思っていました。

ありがとうございます！

Reply:
参照するカラムが一つの場合だと、同じsqlなんですね🤨
```
>> User.first
  User Load (0.9ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> #<User id: 9, name: "a", introduction: "aaa", external_blog_url: "", created_at: "2024-03-29 03:05:14.203009000 +0000", updated_at: "2024-04-18 05:42:31.752793000 +0000">
>> User.first.id
  User Load (0.9ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> 9
```